### PR TITLE
[MODINREACH-280] When Moving a Local Hold Request from One Item to Anot…

### PR DESCRIPTION
**Pull request by aleksandr2110**
…her, Update Local Hold Transaction to TRANSFER State

## Purpose
INN-Reach Local Hold transactions should be updated to TRANSFER state

Requirements/Scope:

Update kafka event handling for Request records to also update Local Hold transactions

Acceptance criteria:
Local Hold transactions are updated when the associated FOLIO request record is moved to a new item

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by oliinyko** [comment]
Check the implementation of transfer Item Hold, the logic there is very similar to what needs to be done here.

**Comment by aleksandr2110** [comment]
> 

Do you mean this method ?
image

**Comment by oliinyko** [comment]
> > 
> 
> Do you mean this method ? image

As well as a method that handles kafka event for moved request.

FYI there is a way to link code snippets in github instead of pasting screenshots...
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-a-permanent-link-to-a-code-snippet

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java [hunk 1]
```diff
@@ -2012,15 +2008,12 @@ void transferLocalHoldItem() {
...
     when(inventoryClient.findItem(item.getId())).thenReturn(Optional.of(item));
     when(circulationClient.findRequest(request.getId())).thenReturn(Optional.of(request));
-    when(circulationClient.moveRequest(eq(request.getId()), any())).thenReturn(request);
 
     var responseEntity = testRestTemplate.postForEntity(
       LOCAL_HOLD_TRANSFER_ITEM_ENDPOINT, null, InnReachTransactionDTO.class,
       PRE_POPULATED_LOCAL_HOLD_TRANSACTION_ID, item.getId()
     );
 
-    verify(requestService).moveItemRequest(request.getId(), item);
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java:2022]: why is it deleted?

Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java:2022]: Fixed

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 2]
```diff
@@ -124,6 +127,9 @@ class KafkaCirculationEventListenerApiTest extends BaseKafkaApiTest {
   @MockBean
   private CirculationClient circulationClient;
 
+  @SpyBean
+  private RequestService requestService;
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: where is it used?

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java [hunk 3]
```diff
@@ -2001,35 +1997,69 @@ void testItemHoldFinalCheckInWhenLoanIsOpen(TransactionState state) {
...
+    item.setStatus(AVAILABLE);
+    var itemId = item.getId();
+    var request = createRequestDTO();
+    request.setId(PRE_POPULATED_LOCAL_HOLD_REQUEST_ID);
+
+    modifyTransactionState(PRE_POPULATED_LOCAL_HOLD_TRANSACTION_ID, LOCAL_HOLD);
+
+    when(inventoryClient.findItem(itemId)).thenReturn(Optional.of(item));
+    when(circulationClient.findRequest(request.getId())).thenReturn(Optional.of(request));
+    when(circulationClient.queryRequestsByItemId(item.getId())).thenReturn((getOpenRequests()));
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java]: this mock is not needed

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 4]
```diff
@@ -637,6 +627,33 @@ private void cancelPatronHoldWithClosedRequest(InnReachTransaction transaction)
...
 
+  private void updateLocalTransactionOnRequestChange(RequestDTO request, InnReachTransaction transaction) {
+    var hold = transaction.getHold();
+    var transactionItemId = hold.getFolioItemId();
+    var itemId = request.getItemId();
+    var requestId = hold.getFolioRequestId();
+
+    if (!transactionItemId.equals(request.getItemId())) {
+      log.info("Updating local hold transaction {} on moving a request {} from item {} to {}",
+        transaction.getId(), requestId, transactionItemId, itemId);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java:638]: you already defined a variable for `transaction.getId()`, reuse it

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java (hunk 2 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: Deleted

File: src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java (hunk 3 above)
Comment (aleksandr2110) [reply src/test/java/org/folio/innreach/controller/InnReachTransactionControllerTest.java]: Deleted

File: src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java [hunk 5]
```diff
@@ -97,6 +97,8 @@ class KafkaCirculationEventListenerApiTest extends BaseKafkaApiTest {
   private static final UUID PRE_POPULATED_ITEM_TRANSACTION_ID = UUID.fromString("ab2393a1-acc4-4849-82ac-8cc0c37339e1");
   private static final UUID PRE_POPULATED_ITEM_TRANSACTION_LOAN_ID = UUID.fromString("06e820e3-71a0-455e-8c73-3963aea677d4");
   private static final UUID PRE_POPULATED_LOCAL_TRANSACTION_ID = UUID.fromString("79b0a1fb-55be-4e55-9d84-01303aaec1ce");
+  private static final UUID PRE_POPULATED_LOCAL_TRANSACTION_REQUEST_ID = UUID.fromString("4106d147-9085-4dfa-a59f-b8d50d551a48");
+  private static final UUID PRE_POPULATED_REQUEST_ITEM_ID = UUID.fromString("b6809be0-8b9b-4407-85a9-ffa4e95a64c9");
```
Comment (oliinyko) [inline src/test/java/org/folio/innreach/domain/listener/KafkaCirculationEventListenerApiTest.java]: is it actually "pre populated"? I dont see it anywhere in our db scripts

**Review by oliinyko — APPROVED** [review]

**Pull request merged**